libkrun/libkruntime.c

```diff
@@ -342,126 +419,213 @@ libkruntime_done(void)
+ * Since some languages cannot represent a uint64_t, we sometimes have to pass
+ * around a double. This is annoying since precision could be silently lost.
+ * This function makes loss of precision explicit, stopping the VM.
  *
- * In the future, a 'shift' argument may be necessary, should we ever read a
- * value from an MSR which doesn't sit flush with bit 0.
+ * We don't expect to actually see a crash since the TSR is zeroed at reboot,
+ * and our benchmarks are not long enough running to generate a large enough
+ * TSR value to cause precision loss (you would need a integer that would not
+ * fit in a 52-bit unsigned int before precision starts being lost in
+ * lower-order bits). Nevertheless, we check it.
+ *
+ * This routine comes at a small cost (a handful of asm instrs). Note that this
+ * cost is a drop in the ocean compared to benchmark workloads.
  */
-uint64_t
-sum_msr_all_cores(long msr_addr, uint64_t mask)
+double
+krun_u64_to_double(uint64_t u64_val)
+{
+    double d_val = (double) u64_val;
+    uint64_t u64_val2 = (uint64_t) d_val;
+
+    if (u64_val != u64_val2) {
+        fprintf(stderr, "Loss of precision detected!\n");
+        fprintf(stderr, "%" PRIu64 " != %" PRIu64 "\n", u64_val, u64_val2);
+        exit(EXIT_FAILURE);
+    }
+    // (otherwise all is well)
+    return d_val;
+}
+
+/*
+ * Load up the results struct.
+ */
+void
+krun_measure(int mdata_idx)
 {
-    int core;
-    uint64_t sum = 0, val;
+    struct krun_data *data = &(krun_mdata[mdata_idx]);
 
-    for (core = 0; core < num_msr_nodes; core++) {
-        val = read_msr(core, msr_addr) & mask;
+    KDATA_BOUNDS_CHECK(mdata_idx);
 
-        // Check for overflow
-        if (sum > UINT64_MAX - val) {
-            fprintf(stderr, "Sum overflow!: "
-                    "%" PRIu64 " + %" PRIu64 "\n", sum, val);
-            exit(EXIT_FAILURE);
+#if defined(__linux__) && !defined(TRAVIS)
+    // We support per-core readings
+
+    // We do it this way to ensure correct nesting of reads
+    if (mdata_idx == 0) {
+        // start readings
+        for (int core = 0; core < krun_num_cores; core++) {
+            data->aperf[core] = krun_read_aperf(core);
+            data->mperf[core] = krun_read_mperf(core);
+            data->core_cycles[core] = krun_read_core_cycles(core);
+        }
+        data->wallclock = krun_clock_gettime_monotonic();
+    } else {
+        // stop readings
+        data->wallclock = krun_clock_gettime_monotonic();
+        for (int core = 0; core < krun_num_cores; core++) {
+            data->core_cycles[core] = krun_read_core_cycles(core);
+            data->aperf[core] = krun_read_aperf(core);
+            data->mperf[core] = krun_read_mperf(core);
         }
-        sum += val;
+        krun_check_mdata();
     }
-
-    return sum;
-}
 #elif defined(__linux__) && defined(TRAVIS)
-// function not used on non-linux or travis
+    data->wallclock = krun_clock_gettime_monotonic();
+    if (mdata_idx == 1) {
```

Comment:
Why is the `if` inside a platform check? Shouldn't it be always executed and thus outside the platform check?